### PR TITLE
Post: New blog post about performance boost in Hoa\Compiler

### PR DESCRIPTION
**Pull request by Hywan**
Please @hoaproject/hoackers, can you review this blog post :smiley:?

### CODE REVIEW

File: Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+      (<a href="http://hoa-project.net/Literature/Hack/Compiler.html">Learn more
+      by reading its documentation</a>). Until today.
+    </p>
+    <p>
+      The issue with this approach is that if the datum is of length 1Mb, then
+      the memory will contain 1Mb (the original datum) and at least 1Mb (the
+      sequence). However, the sequence is represented as a PHP array, so it has
+      an overhead of some bytes per tokens, in addition to the meta data attached
+      to the token (like the name, the namespace, the length, the offset etc.).
+      The resulting overhead size will depend of the PHP version in use but this
```
Comment (jubianchi) [inline Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl:34]: "will depend on"

File: Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+      is not optimal.
+    </p>
+    <p>
+      Thus, a straightforward optimisation consists at transforming the lexer into
```
Comment (jubianchi) [inline Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl:38]: `s/consists at transforming/is to transform/`

File: Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+      an iterator. Actually, it has been <strong>transformed into a
+      generator</strong>. So each call to the lexer will produce the next
+      token, without stacking all the tokens in memory.
+    </p>
+    <p>
+      With this new approach, the API stays <strong>identically the
```
Comment (jubianchi) [inline Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl:44]: "identically the same" seems repetitive. Why not just use either "identical" or "the same" ?

File: Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    </figure>
+
+    <h2>Save memory and CPU in the parser, and pragmas</h2>
+
+    <p>
+      Now, the parser takes the benefits from the lexer since it is an iterator.
+      Method calls have been reduced, so many CPU cycles have been saved. Also,
+      <strong>indirections have been reduced</strong>. When a datum points to
+      another datum that points to the final one, we count 2 indirections to get
+      the final datum. A redirection has a cost. This cost has been reduced.
```
Comment (jubianchi) [inline Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl:69]: "redirection" ? shloudn't it be "indirection" ?

File: Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl [hunk 5]
```diff
@@ -1,16 +1,18 @@
...
   <article id="main">
     <p>
       A lot of work has been done <strong>to boost the performance</strong> of
       <a href="http://central.hoa-project.net/Resource/Library/Compiler"><code>Hoa\Compiler</code></a>,
-      mainly by reducing the amount of memory required to lexically or to
-      syntactically analyse it. With a large data, <strong>up to 80% of the
-       memory can be saved</strong>.
+      mainly by reducing the amount of memory required to lexically and to
+      syntactically analyse a data. <strong>Up to 80% of the
+      memory can be saved</strong> with a large data. Also a new feature allows
```
Comment (CircleCode) [inline Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl:13]: I think you can't say _**a** large data_, but rather _large data_

File: Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+      Another pragma has been introduced to disable Unicode support in the
+      lexer:
+    </p>
+    <pre><code>%pragma lexer.unicode false</code></pre>
+    <p>
+      This is particularly useful when the grammar defines a binary language or
+      defines its own Unicode support. This is the case of
+      <a href="http://central.hoa-project.net/Resource/Library/Json"><code>Hoa\Json</code></a>.
+      The RFC7159 is under fully implementation and JSON uses the same UTF-8
+      format than Javascript: With surrogate pairs. This was not possible to
```
Comment (jubianchi) [inline Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl:142]: `s/This was not/It was no/`

File: Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+      correctly lexically analyse a JSON string with PCRE Unicode support
+      enabled. With this new pragma, this is possible.
```
Comment (jubianchi) [inline Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl:144]: `s/this is possible/it is possible/`

File: Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+<?xyl-meta name="date" value="2016-08-22T08:17:28+02:00"?>
+
+<overlay xmlns="http://hoa-project.net/xyl/xylophone">
+  <article id="main">
+    <p>
+      A lot of work has been done <strong>to boost the performance</strong> of
+      <a href="http://central.hoa-project.net/Resource/Library/Compiler"><code>Hoa\Compiler</code></a>,
+      mainly by reducing the amount of memory required to lexically and to
+      syntactically analyse a data. <strong>Up to 80% of the
+      memory can be saved</strong> with a large data. Also a new feature allows
```
Comment (CircleCode) [inline Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl:13]: no _**a** large data_ but _large data_

File: Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl (hunk 5 above)
Comment (jubianchi) [reply Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl:13]: and use "datum" as it's done in the rest of the post

File: Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+      Thus, a straightforward optimisation consists at transforming the lexer into
+      an iterator. Actually, it has been <strong>transformed into a
+      generator</strong>. So each call to the lexer will produce the next
+      token, without stacking all the tokens in memory.
+    </p>
+    <p>
+      With this new approach, the API stays <strong>identically the
+      same</strong> and the memory peak usage is <strong>drastically
+      reduced</strong>.
+    </p>
```
Comment (CircleCode) [inline Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl:47]: does it still allows look forward and look behind?

File: Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl (hunk 3 above)
Comment (Hywan) [reply Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl:44]: :+1:

File: Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl (hunk 2 above)
Comment (Hywan) [reply Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl:38]: :+1:

File: Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl (hunk 1 above)
Comment (Hywan) [reply Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl:34]: :+1:

File: Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl (hunk 4 above)
Comment (Hywan) [reply Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl:69]: Of course. :+1:

File: Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl (hunk 5 above)
Comment (Hywan) [reply Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl:13]: :+1:

File: Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl (hunk 6 above)
Comment (Hywan) [reply Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl:142]: :+1:

File: Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl (hunk 8 above)
Comment (Hywan) [reply Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl:13]: :+1:

File: Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl (hunk 9 above)
Comment (Hywan) [reply Source/Posts/2016/08-Performance-boost-for-Hoa-Compiler.xyl:47]: Yup it does!

**Comment by Hywan** [comment]
Published, http://blog.hoa-project.net/2016/08-Performance-boost-for-Hoa-Compiler.html \o/. Thanks for the reviews!

**Pull request merged**